### PR TITLE
Compatibility with Mocha's parallel mode

### LINKED ISSUE #619: Mocha parallel mode and Allure API
**Describe the bug**
We have a problem related to mocha parallel mode for a long time, there is some related issues in the repository: #485, #486.

**To Reproduce**
1. Use allure mocha reporter in parallel mode
2. Try to call any allure API method such as label
3. You'll see an error in console

**Expected behavior**
We should be able to use Allure API with mocha in parallel mode.

**Additional context**
At this moment, the problem can't be solved because mocha parallel mode implementation. The running uses `workerpool` module, what doesn't work with plain node IPC messages. Additionally, mocha doesn't provide any way to send messages from tests to the reporter, what works in the main thread, so we can't do anything at this moment to make both parts work correctly.

As temporal solution, I suggest to use single thread mode, if you need Allure API.

All the updates about the problem I'll post here, in the issue.

### PR DESCRIPTION
**Pull request by delatrie**
### Context

The PR allows Allure Mocha to be used with Mocha's `--parallel` mode. It does so by hooking into Mocha's `ParallelBuffered` reporter's [`createListeners`](https://github.com/mochajs/mocha/blob/99601da68d59572b6aa931e9416002bcb5b3e19d/lib/nodejs/reporters/parallel-buffered.js#L119) in a worker thread. Not perfect, but, you know, lesser of two evils.

The patch is set up by introducing a new module into the list of modules Mocha requires at runtime (via `Mocha.MochaOptions.require`). The module only affects worker processes.

#### New ways to pass a writer into the runtime

A writer can now be provided as a string or an array. That makes it easier to pass the config options across process boundaries.

Now, the following ways are supported:

1. An instance (previously, that was the only way).
2. A string containing a well-known class name or the full path to a module.
3. An array of one or more items. The first one must be a well-known class name or the full path to a module, while the rest are the arguments to the constructor the the class.

The list of well-known classes is as follows:

  - `AllureInMemoryAllureWriter` (all environments)
  - `FileSystemAllureWriter` (node.js only)
  - `AllureInMemoryWriter` (node.js only)
  - `MessageAllureWriter` (node.js only)

The module must export either a writer class or an instance.

Currently, only CJS modules are supported. To support ESM modules, an async extension point is required somewhere in `allure-js-commons.` Since that's primarily a mechanism to test Allure integrations, we might not urgently need it.

Fixes #619.

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure-js

### CODE REVIEW

**Review by epszaw — CHANGES_REQUESTED** [review]
I'm not able to compile the module due the following error:

```shell
λ yarn compile

src/index.ts → dist/index.mjs...
src/setupAllureMochaParallel.ts:3:45 - error TS7016: Could not find a declaration file for module 'mocha/lib/nodejs/reporters/parallel-buffered.js'. '/Users/epszaw/Code/Work/qameta/allure-js/.yarn/cache/mocha-npm-10.3.0-2a85896b70-8b30b3fdb3.zip/node_modules/mocha/lib/nodejs/reporters/parallel-buffered.js' implicitly has an 'any' type.
  If the 'mocha' package actually exposes this module, consider sending a pull request to amend 'https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/mocha'

3 import { default as ParallelBuffered } from "mocha/lib/nodejs/reporters/parallel-buffered.js";
                                              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in src/setupAllureMochaParallel.ts:3

ERROR: "compile:types" exited with 2.
```

File: packages/allure-mocha/rollup.config.mjs [hunk 1]
```diff
@@ -53,5 +55,5 @@ const createNodeEntry = (inputFile) => {
 };
 
 export default () => {
-  return [createNodeEntry("src/index.ts")].flat();
```
[the remark this answers is not in the document]
Comment (delatrie) [reply packages/allure-mocha/rollup.config.mjs:56]: It's for Mocha, not end users. We need Mocha to require this file in its workers via `MochaOptions.requires`. That's why we need it to be a separate file, not part of the bundle.

File: packages/allure-mocha/src/utils.ts [hunk 2]
```diff
@@ -15,3 +19,26 @@ export const errorToStatusDetails = (error: unknown): StatusDetails | undefined
 };
 
 export const getSuitesOfMochaTest = (test: Mocha.Test) => test.titlePath().slice(0, -1);
+
+export const resolveParallelModeSetupFile = () => join(__dirname, `setupAllureMochaParallel${extname(__filename)}`);
```
[the remark this answers is not in the document]
Comment (delatrie) [reply packages/allure-mocha/src/utils.ts:23]: Mocha, not us, should require it. Once the full path to that file is put in `MochaOptions.requires`, Mocha's workers will require it (similar to if it was provided via the [`--require`](https://mochajs.org/#-require-module-r-module) CLI argument)

File: packages/allure-js-commons/src/sdk/utils.ts [hunk 3]
```diff
@@ -177,3 +179,30 @@ export const readImageAsBase64 = async (filePath: string): Promise<string | unde
     return undefined;
   }
 };
+
+export type WellKnownWriters = {
```
[the remark this answers is not in the document]
Comment (delatrie) [reply packages/allure-js-commons/src/sdk/utils.ts]: Yep, sure. Done.

**Review by epszaw — APPROVED** [review]

**Pull request merged**